Add widget type endpoint.

spacedmonkey opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Add widget type endpoint. 

To test. 

`__experimental/widget-types`
`__experimental/widget-types/<name>`

Fixes: #25969 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

spacedmonkey committed:
Add widget type endpoint.

TimothyBJacobs commented on lib/class-wp-rest-widget-types-controller.php:
```diff
@@ -0,0 +1,456 @@
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<name>[a-zA-Z0-9_-]+)',
```
Maybe make this available at `/widget/<name>/form-renderer` to be similar to `block-renderer`? I think it just being the `POST` request at the widget endpoint itself is a bit too broad.

spacedmonkey replied:
I had it with way before, change to this. But you are right. I changed it back. https://github.com/WordPress/gutenberg/pull/26042/commits/7d3884e82dbe50581da253ebb95580046e096d20

TimothyBJacobs commented on lib/class-wp-rest-widget-types-controller.php:
```diff
@@ -0,0 +1,456 @@
+		global $wp_widget_factory;
+
+		$data = array();
+		foreach ( $wp_widget_factory->widgets as $slug => $widget ) {
```
This needs to support widgets registered with `wp_register_sidebar_widget()`. See `::do_test_update_item_legacy_widget`.

spacedmonkey replied:
I don't this it needs to. This is a endpoint for widget types. `wp_register_sidebar_widget` does not register a widget type, it registers a instance of a widget. Unless I create this code wrong.

TimothyBJacobs replied:
Then where does the `description` live? Will we duplicate that field in the widgets controller?

spacedmonkey replied:
It lives here. https://github.com/WordPress/gutenberg/pull/26042/files#diff-a6036d6ccb63c98c7f607f1bcc572b84a185f194e3668d747f37123ba1207e60R370

And yes, it will be repeated fields. Which is why I wanted to add this now, so we can remove those fields from the widget controller.

TimothyBJacobs replied:
That's not what I'm saying. Where will the `description` field for widgets registered with `wp_register_sidebar_widget` live?

If we remove the repeated fields from the widgets controller, then there will be no way to get legacy widget's descriptions since they won't have an entry in the types controller.

spacedmonkey replied:
Oh, yeah. I guess we would have to leave that in the widget controller.

TimothyBJacobs replied:
This would also effect all of the other arguments for `widget_options` as well as for controls registered with `wp_register_widget_control`. So I don't see how we can meaningfully implement a separate types controller without support for legacy widgets as long as we are going to support legacy widgets at all.

TimothyBJacobs replied:
I think we should be able to iterate over the ones registered with `wp_register_sidebar_widget` and exclude ones that have an id base that correspond to an entry in the widgets factory.

spacedmonkey replied:
I don't see widget registered with `wp_register_sidebar_widget` as a valid widget type. It is setup up a instance of a widget, not a widget type. There are two different things. I don't think they should be support here, as it does make any sense and the fields do not match at all. WP_Widget has been around for a LONG time, since 2.8.0. I am also not overly worried about `wp_register_sidebar_widget`. Anyone using a widget like that has likely not upgraded.

TimothyBJacobs replied:
Gutenberg currently supports it both in the UI and the server. Dropping support for that would have to be done in a separate issue.

I'm also not sure how the fields are incompatible, they should be identical. `WP_Widget::_register_one` registers itself as a widget with the defined widget options in the constructor.

And the `control_options` in the `WP_Widget` constructor are defined by referencing the ones supported by `wp_register_widget_control`.

spacedmonkey replied:
Fixed in https://github.com/WordPress/gutenberg/pull/26042/commits/f18fc367184fbb98b25089034528b487cc88f2ad

TimothyBJacobs replied:
This is great, thank you! Could we add a test copying the widget definition from `do_test_update_item_legacy_widget`?

TimothyBJacobs replied:
Sorry, totally missed you added that already.

TimothyBJacobs commented on phpunit/class-rest-widget-types-controller-test.php:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	/**
+	 * @expectedIncorrectUsage rest_sanitize_value_from_schema
```
What is causing us to issue `_doing_it_wrong` warnings?

spacedmonkey replied:
I didn't work that out, was going to look at this tommorrow.

spacedmonkey replied:
Worked it out.

TimothyBJacobs commented on lib/class-wp-rest-widget-types-controller.php:
```diff
@@ -0,0 +1,456 @@
+				),
+				'control_options' => array(
+					'description'          => __( 'Control options', 'gutenberg' ),
+					'type'                 => 'Object',
```
```suggestion
					'type'                 => 'object',
```

spacedmonkey replied:
Fixed https://github.com/WordPress/gutenberg/pull/26042/commits/7d3884e82dbe50581da253ebb95580046e096d20

TimothyBJacobs commented on lib/class-wp-rest-widget-types-controller.php:
```diff
@@ -0,0 +1,456 @@
+				),
+				'widget_options'  => array(
+					'description'          => __( 'Widget options', 'gutenberg' ),
+					'type'                 => 'Object',
```
```suggestion
					'type'                 => 'object',
```

spacedmonkey replied:
Fixed https://github.com/WordPress/gutenberg/pull/26042/commits/7d3884e82dbe50581da253ebb95580046e096d20

TimothyBJacobs commented on lib/class-wp-rest-widget-types-controller.php:
```diff
@@ -0,0 +1,456 @@
+						),
+					),
+					'additionalProperties' => array(
+						'type' => 'Object',
```
```suggestion
						'type' => 'object',
```

Also, are we sure this is accurate? Why can't additional args be of any type?

spacedmonkey replied:
Fixed https://github.com/WordPress/gutenberg/pull/26042/commits/7d3884e82dbe50581da253ebb95580046e096d20

spacedmonkey committed:
Feedback

spacedmonkey commented:
@TimothyBJacobs I would like to get this merged before #25232 . From there I will rebase the PR and add links into #25232

TimothyBJacobs commented:
We can merge this before #25232, but that PR should stay the same so that the client can implement changes after it is merged.

spacedmonkey committed:
Remove incorrect usage call.

spacedmonkey committed:
Support legecy widget types

spacedmonkey committed:
Add test for legacy.

spacedmonkey committed:
Add a test.

TimothyBJacobs committed:
Switch to wp/v2 namespace.

This is either going to land in Gutenberg 9.2.0 for release in 5.6 which means
it has to be stable anyway, or if that doesn't happen we can revert.

TimothyBJacobs committed:
Merge branch 'master' into feature/widget-type

TimothyBJacobs committed:
Syntax error

TimothyBJacobs reviewed (approved):
Looks good, let's get this in so the client can start adjusting.

talldan commented:
Hey @TimothyBJacobs and @spacedmonkey, just seeing some failing tests in `master`:
```
1) REST_Widget_Types_Controller_Test::test_get_items
87
Failed asserting that actual size 18 matches expected size 20.
```

Not sure how this one materialized, as I can't see the error in this PR.

talldan commented:
Thanks for the fix 🙇
